### ISSUE TITLE
GCC warnings if [-Wshadow] is enabled. 

### ISSUE THREAD

**Issue by Remotion** [fmtlib/fmt#770]
```
In file included from \test\warnings_test.cpp:3:0:
fmt/format.h: In function 'Char* fmt::v5::internal::format_decimal(Char*, UInt, unsigned int, ThousandsSep)':
fmt/format.h:1081:55: warning: declaration of 'buffer' shadows a global declaration [-Wshadow]
                             ThousandsSep thousands_sep) {
                                                       ^
In file included from fmt/format.h:67:0,
                 from \test\warnings_test.cpp:3:
fmt/core.h:390:28: note: shadowed declaration is here
 typedef basic_buffer<char> buffer;
                            ^~~~~~
In file included from \test\warnings_test.cpp:3:0:
fmt/format.h: In function 'Iterator fmt::v5::internal::format_decimal(Iterator, UInt, unsigned int, ThousandsSep)':
fmt/format.h:1111:59: warning: declaration of 'buffer' shadows a global declaration [-Wshadow]
   char_type buffer[std::numeric_limits<UInt>::digits10 + 2];
                                                           ^
In file included from fmt/format.h:67:0,
                 from \test\warnings_test.cpp:3:
fmt/core.h:390:28: note: shadowed declaration is here
 typedef basic_buffer<char> buffer;
                            ^~~~~~
In file included from \test\warnings_test.cpp:3:0:
fmt/format.h: In function 'Char* fmt::v5::internal::format_uint(Char*, UInt, unsigned int, bool)':
fmt/format.h:1123:44: warning: declaration of 'buffer' shadows a global declaration [-Wshadow]
                          bool upper = false) {
                                            ^
In file included from fmt/format.h:67:0,
                 from \test\warnings_test.cpp:3:
fmt/core.h:390:28: note: shadowed declaration is here
 typedef basic_buffer<char> buffer;
                            ^~~~~~
In file included from \test\warnings_test.cpp:3:0:
fmt/format.h: In function 'It fmt::v5::internal::format_uint(It, UInt, unsigned int, bool)':
fmt/format.h:1139:64: warning: declaration of 'buffer' shadows a global declaration [-Wshadow]
   char buffer[std::numeric_limits<UInt>::digits / BASE_BITS + 2];
                                                                ^
In file included from fmt/format.h:67:0,
                 from \test\warnings_test.cpp:3:
fmt/core.h:390:28: note: shadowed declaration is here
 typedef basic_buffer<char> buffer;
                            ^~~~~~
In file included from \test\warnings_test.cpp:3:0:
fmt/format.h: In constructor 'constexpr fmt::v5::internal::arg_ref<Char>::arg_ref(unsigned int)':
fmt/format.h:1913:50: warning: declaration of 'index' shadows a member of 'fmt::v5::internal::arg_ref<Char>' [-Wshadow]
   FMT_CONSTEXPR explicit arg_ref(unsigned index) : kind(INDEX), index(index) {}
                                                  ^
fmt/format.h:1924:14: note: shadowed declaration is here
     unsigned index;
              ^~~~~
fmt/format.h: In constructor 'fmt::v5::internal::arg_ref<Char>::arg_ref(fmt::v5::basic_string_view<Char>)':
fmt/format.h:1914:50: warning: declaration of 'name' shadows a member of 'fmt::v5::internal::arg_ref<Char>' [-Wshadow]
   explicit arg_ref(basic_string_view<Char> name) : kind(NAME), name(name) {}
                                                  ^
fmt/format.h:1925:29: note: shadowed declaration is here
     basic_string_view<Char> name;
                             ^~~~
fmt/format.h: In member function 'constexpr typename ParseContext::iterator fmt::v5::formatter<T, Char, typename std::enable_if<fmt::v5::internal::format_type<typename fmt::v5::buffer_context<Char>::type, T>::value>::type>::parse(ParseContext&)':
fmt/format.h:3175:10: warning: declaration of 'type_spec' shadows a global declaration [-Wshadow]
     auto type_spec = specs_.type();
          ^~~~~~~~~
fmt/format.h:1275:37: note: shadowed declaration is here
 typedef format_spec<char, type_tag> type_spec;
                                     ^~~~~~~~~
fmt/format.h: In constructor 'fmt::v5::arg_join<It, Char>::arg_join(It, It, fmt::v5::basic_string_view<Char>)':
fmt/format.h:3405:5: warning: declaration of 'sep' shadows a member of 'fmt::v5::arg_join<It, Char>' [-Wshadow]
     : begin(begin), end(end), sep(sep) {}
     ^
fmt/format.h:3402:27: note: shadowed declaration is here
   basic_string_view<Char> sep;
                           ^~~
fmt/format.h:3405:5: warning: declaration of 'end' shadows a member of 'fmt::v5::arg_join<It, Char>' [-Wshadow]
     : begin(begin), end(end), sep(sep) {}
     ^
fmt/format.h:3401:6: note: shadowed declaration is here
   It end;
      ^~~
fmt/format.h:3405:5: warning: declaration of 'begin' shadows a member of 'fmt::v5::arg_join<It, Char>' [-Wshadow]
     : begin(begin), end(end), sep(sep) {}
     ^
fmt/format.h:3400:6: note: shadowed declaration is here
   It begin;
      ^~~~~
In file included from fmt/format.h:3736:0,
                 from \test\warnings_test.cpp:3:
fmt/format-inl.h: In static member function 'static int fmt::v5::internal::char_traits<char>::format_float(char*, std::size_t, const char*, int, T)':
fmt/format-inl.h:226:79: warning: declaration of 'buffer' shadows a global declaration [-Wshadow]
     char *buffer, std::size_t size, const char *format, int precision, T value) {
                                                                               ^
In file included from fmt/format.h:67:0,
                 from \test\warnings_test.cpp:3:
fmt/core.h:390:28: note: shadowed declaration is here
 typedef basic_buffer<char> buffer;
                            ^~~~~~
In file included from fmt/format.h:3736:0,
                 from \test\warnings_test.cpp:3:
fmt/format-inl.h: In static member function 'static int fmt::v5::internal::char_traits<wchar_t>::format_float(wchar_t*, std::size_t, const wchar_t*, int, T)':
fmt/format-inl.h:235:12: warning: declaration of 'buffer' shadows a global declaration [-Wshadow]
     T value) {
            ^
In file included from fmt/format.h:67:0,
                 from \test\warnings_test.cpp:3:
fmt/core.h:390:28: note: shadowed declaration is here
 typedef basic_buffer<char> buffer;
```

**Comment by eliaskosunen** [comment]
I assume you're only including <core.h>, because `-Wshadow` is explicitly disabled here:
https://github.com/fmtlib/fmt/blob/1b8a7f8fa0a3c709984af9a5ade08a0d39196260/include/fmt/format.h#L54-L65

**Comment by Remotion** [comment]
No.
```cpp
#define FMT_HEADER_ONLY
#include "fmt/format.h"

int main() {
    std::string ss = "string";
    fmt::format("{} \n", ss);

    return 0;
}
```

**Comment by eliaskosunen** [comment]
Can't reproduce in CE: https://godbolt.org/g/gqstpW
What version on gcc are you using and are you sure that the version of fmt is post-`691a7a91a1d7cf65c75810adeebb737c38db9993`

**Comment by Remotion** [comment]
GCC 7.3 from here  https://nuwen.net/mingw.html

g++  -std=c++1z -pipe -Wall -Wextra -Wshadow -Wreturn-type -fconcepts -fstrong-eval-order -faligned-new -march=native

{fmt} is 1b8a7f8fa0a3c709984af9a5ade08a0d39196260 commit.

**Comment by eliaskosunen** [comment]
I think this might be an issue with MinGW rather than fmt, as I still can't reproduce with gcc 7.3 and with the same compiler flags: https://godbolt.org/g/tnNB8T

Does anyone have gcc 7.3 installed locally so they can verify?

What values are `__GNUC__` and `__clang__` defined to in your platform when compiling with those settings?

**Comment by Remotion** [comment]
__GNUC__ is 7 and __clang__ is not defined.

**Comment by eliaskosunen** [comment]
You're correct, that comparison is incorrect. The proper expression would be `(__GNUC__ * 100 + __GNUC_MINOR__) >= 406`. You could probably submit a PR.

**Comment by vitaut** [comment]
Fixed in https://github.com/fmtlib/fmt/commit/47268ecd808b86eb35b617d3db3078daf4641fcc.

**Comment by vitaut** [comment]
AFAICS shadowing warnings are suppressed.

**Comment by 0x8000-0000** [comment]
/usr/bin/g++   -I/home/florin/work/geometry/external/lib/fmt/include -Wall -Wextra -Werror -pedantic -Wconversion -Wunused -Wshadow -g   -std=c++17 -MD -MT external/lib/fmt/CMakeFiles/fmt.dir/src/format.cc.o -MF external/lib/fmt/CMakeFiles/fmt.dir/src/format.cc.o.d -o external/lib/fmt/CMakeFiles/fmt.dir/src/format.cc.o -c /home/florin/work/geometry/external/lib/fmt/src/format.cc


Will produce the warnings above, with GCC 8.2

**Comment by vitaut** [comment]
I see, the suppression doesn't apply in that case. Should be fixed in https://github.com/fmtlib/fmt/commit/fa1d4dbcfd8d2027ac138ba2cafb5d9262904e5c.

**Comment by 0x8000-0000** [comment]
> I see, the suppression doesn't apply in that case. Should be fixed in [fa1d4db](https://github.com/fmtlib/fmt/commit/fa1d4dbcfd8d2027ac138ba2cafb5d9262904e5c).

Confirmed. Thank you!
